### ISSUE TITLE
In language test support

### ISSUE THREAD

**Issue by smoothdeveloper** [fsharp/fslang-suggestions#1133]
This is a broader suggestion related to testing and software quality, right in the core of the language, than a specific feature.

**in language test support:**

As prior art, languages such as Cobra and Rust support concept of "test blocks":

* Cobra: http://cobra-language.com/ http://cobra-language.com/trac/cobra/wiki/Tests 
* Rust: https://www.rust-lang.org/ https://doc.rust-lang.org/book/ch11-01-writing-tests.html

**design by contract**

* Effeil (https://www.eiffel.org/) and Cobra implements "design by contract" with `require`and `ensure` blocks, which are great way to encode invariants in asserts (I assume the constructs may be useful for "logical" linear type analyzer kind of stuff, that could parse those blocks and do flow analysis).

**verifiable interpreter output**

See doctest in Python and Haskell:
* https://docs.python.org/3/library/doctest.html
* https://serokell.io/blog/doctests-in-haskell

For this, I'd actually prefer we devise something which refer to free standing .fsx and their "golden output", a bit like some of our baseline driven test infrastructure.

Having all those things, baked in the language, may have a significant impact on the quality of codebases built on it, by establishing clear guidelines, to make important code humanely maintainable where it matters.

**more ideas**

open for people to share more things they've come across or thought of, in context of language test support.

# Pros

* zeroes cost of context switch, with things like picking a test runner, creating and managing more fsproj in a solution
* F# join the set languages where this is "done right"
* more libraries with more invariants that may invariably hold

# Cons

* tons of work, especially to integrate in comprehensive API in FCS
* people may not embrace those specific approaches as being worthy of specific attention in the language

# Extra information

Estimated cost:  XXL

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [ ] I or my company would be willing to help implement and/or test this

**Comment by baronfel** [comment]
@charlesroddie you're basically describing Expecto. It supports tests as first class values, and has a dotnet test adapter to allow for test running in a unified way. Basically all of my projects have been using it for years specifically for the organizational benefits you outline.

**Comment by charlesroddie** [comment]
Thanks for the suggestion @baronfel . I am exploring YoloDev.Expecto.Sdk but so far this doesn't allow a good connection to test explorer (folders hierarchy does not match up) https://github.com/haf/expecto/issues/431 .

**Comment by dsyme** [comment]
@charlesroddie I'm labelling this as "probably not" because it's just very difficult to get into the test-framework level when things like Expecto, NUnit, XUnit exist, with all their pros and cons.  An in-language solution would effectively compete with one of these but with some of the pros and some of the cons.

**Comment by cartermp** [comment]
Yeah if this were all from scratch I'd actually love a Go-style approach where the convention is just `filename_test`, it lives side-by-side, and the testing tool just picks up those files by convention and works
